pkg/controller/registrysyncer/registrysyncer.go

```diff
@@ -0,0 +1,336 @@
+	r := &reconciler{
+		ctx:              context.Background(),
+		log:              log,
+		registryClients:  map[string]ctrlruntimeclient.Client{},
+		pullSecretGetter: pullSecretGetter,
+	}
+	for clusterName, m := range managers {
+		r.registryClients[clusterName] = imagestreamtagwrapper.MustNew(m.GetClient(), m.GetCache())
+	}
+	c, err := controller.New(ControllerName, mgr, controller.Options{
+		Reconciler: r,
+		// We conflict on ImageStream level which means multiple request for imagestreamtags
+		// of the same imagestream will conflict so stay at one worker in order to reduce the
+		// number of errors we see. If we hit performance issues, we will probably need cluster
+		// and/or imagestream level locking.
+		MaxConcurrentReconciles: 1,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to construct controller: %w", err)
+	}
+
+	for _, m := range managers {
+		if err := c.Watch(
+			source.NewKindWithCache(&imagev1.ImageStream{}, m.GetCache()),
+			handlerFactory(testInputImageStreamTagFilterFactory()),
+		); err != nil {
+			return fmt.Errorf("failed to create watch for ImageStreams: %w", err)
+		}
+	}
+
+	r.log.Info("Successfully added reconciler to manager")
+	return nil
+}
+
+type objectFilter func(types.NamespacedName) bool
+
+// handlerFactory produces a handler that:
+// * Watches ImageStreams because ImageStreamTags do not support the watch verb
+// * Extracts all ImageStramTags out of the Image
+// * Filters out the ones that are not in use
+// Note: We can not use a predicate because that is directly applied on the source and the source yields ImageStreams, not ImageStreamTags
+// * Creates a reconcile.Request per cluster and ImageStreamTag
+func handlerFactory(filter objectFilter) handler.EventHandler {
+	return imagestreamtagmapper.New(func(in reconcile.Request) []reconcile.Request {
+		if !filter(in.NamespacedName) {
+			return nil
+		}
+		return []reconcile.Request{in}
+	})
+}
+
+type reconciler struct {
+	ctx              context.Context
+	log              *logrus.Entry
+	registryClients  map[string]ctrlruntimeclient.Client
+	pullSecretGetter func() []byte
+}
+
+func (r *reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithField("request", req.String())
+	log.Info("Starting reconciliation")
+	err := r.reconcile(req, log)
+	if err != nil {
+		log.WithError(err).Error("Reconciliation failed")
+	} else {
+		log.Info("Finished reconciliation")
+	}
+	return reconcile.Result{}, controllerutil.SwallowIfTerminal(err)
+}
+
+func (r *reconciler) reconcile(req reconcile.Request, log *logrus.Entry) error {
+
+	isTags := map[string]*imagev1.ImageStreamTag{}
+	for clusterName, client := range r.registryClients {
+		imageStreamTag := &imagev1.ImageStreamTag{}
+		if err := client.Get(r.ctx, req.NamespacedName, imageStreamTag); err != nil {
+			if apierrors.IsNotFound(err) {
+				log.Debugf("Source imageStreamTag %s not found on cluster %s", req.NamespacedName.String(), clusterName)
```

Comment:
Please never use `log.Anythingf` and use proper fields instead

Reply:
Updated.
Not sure if i did it right.